### PR TITLE
Adding a way to use decorator instead of set_model for model_as_code

### PR DESCRIPTION
**Pull request by sunishsheth2009**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/sunishsheth2009/mlflow/pull/12911?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12911/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12911
```

</p>
</details>

### What changes are proposed in this pull request?
Adding a way to use decorator instead of set_model for model_as_code

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Adding a way to use decorator instead of set_model for model_as_code

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/models/model.py [hunk 1]
```diff
@@ -1010,22 +1012,68 @@ def set_model(model):
...
-        except Exception as e:
-            raise mlflow.MlflowException(
-                "Model should either be an instance of PyFuncModel or Langchain type."
-            ) from e
-
-    globals()["__mlflow_model__"] = model
+                ) from e
+        globals()["__mlflow_model__"] = model
+
+    # Check if the callable is a function (used as a decorator)
```
Comment (serena-ruan) [inline mlflow/models/model.py:1036]:
```suggestion
    # Check if the callable is a class (used as a decorator)
```

File: mlflow/models/model.py [hunk 2]
```diff
@@ -1010,22 +1012,68 @@ def set_model(model):
...
+        #             "question": RunnablePassthrough(),
+        #         }
+        #         | prompt
+        #         | get_fake_chat_model()
+        #         | StrOutputParser()
+        #     )
+        #   retrieval_chain()
+        def wrapper(*args, **kwargs):
+            returned_model = model(*args, **kwargs)
+            validate_and_set_model(returned_model)
```
Comment (serena-ruan) [inline mlflow/models/model.py:1075]:
```suggestion
            validate_and_set_model(returned_model)
            return returned_model
```

File: mlflow/models/model.py [hunk 3]
```diff
@@ -1010,22 +1012,68 @@ def set_model(model):
     Args:
         model: The model object to be logged.
     """
-    from mlflow.pyfunc import PythonModel
 
-    if not (isinstance(model, PythonModel) or callable(model)):
-        try:
-            from mlflow.langchain import _validate_and_prepare_lc_model_or_path
+    def validate_and_set_model(model):
```
Comment (serena-ruan) [inline mlflow/models/model.py:1016]: Can we move this function outside of set_model?

File: mlflow/models/model.py [hunk 4]
```diff
@@ -1010,22 +1012,68 @@ def set_model(model):
...
-            raise mlflow.MlflowException(
-                "Model should either be an instance of PyFuncModel or Langchain type."
-            ) from e
-
-    globals()["__mlflow_model__"] = model
+                ) from e
+        globals()["__mlflow_model__"] = model
+
+    # Check if the callable is a function (used as a decorator)
+    if inspect.isclass(model):
```
Comment (serena-ruan) [inline mlflow/models/model.py:1037]: In general, could we use a different function for the decorator (e.g. @model_as_code) instead of mixing it with the set_model function? I feel like it's hard to control the behavior and we might miss some corner cases.

**Review by serena-ruan — COMMENTED** [review]
Could we update docs for this change accordingly?

File: tests/pyfunc/sample_code/python_model_decorator.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+from mlflow.pyfunc import PythonModel
+
+
+@set_model
+class MyModel(PythonModel):
+    def predict(self, context, model_input):
+        return f"This was the input: {model_input}"
+
+
+MyModel()
```
Comment (harupy) [inline tests/pyfunc/sample_code/python_model_decorator.py:11]: How is this different from the following code?

```python
class MyModel(PythonModel):
    def predict(self, context, model_input):
        return f"This was the input: {model_input}"


set_model(MyModel())
```

What's the benefit of using `set_model` as a decorator?

Comment (serena-ruan) [reply tests/pyfunc/sample_code/python_model_decorator.py:11]: +1, do customers prefer decorators over direct API call?

Comment (sunishsheth2009) [reply tests/pyfunc/sample_code/python_model_decorator.py:11]: The motivation is that customers are not super familiar with this paradigm of setting a model this way by calling a function.
users are more familiar with decorators now, example for creating a step in langgraph, or annotating a trace. So was wondering if we can create a similar API for the same.

What do you think?
I am down to create a new name like `@model` or something but that module is already exposed in MLflow so can't reuse that :( 
Any name suggestions?

File: mlflow/models/model.py (hunk 4 above)
Comment (sunishsheth2009) [reply mlflow/models/model.py:1037]: Yes I am down to split it, I agree, it would be more cleaner. But I wasn't sure of a good name for this. `model_as_code` might be too big. i was thinking `@model` but thats already a module in MLflow so can't reuse it as far as i know :(.
Any suggestions?

**Comment by BenWilson2** [comment]
@sunishsheth2009 can we solve this with more effective documentation of using this new paradigm? 

https://github.com/mlflow/mlflow/pull/12936 
Having two ways to do the same thing is pretty confusing.

**Comment by harupy** [comment]
+1 for improving the docs.

**Comment by sunishsheth2009** [comment]
Sounds good. Thank you guys, we can see if customers still get confused after the docs improvement and we can revisit this then :)
